### ISSUE TITLE
Run all Puppeteer commands in an Isolated World

### ISSUE THREAD

**Issue by prescience-data** [berstend/puppeteer-extra#224]
Attempting to address this issue below and looking for feedback and ideas on any potential ways to achieve this more cleanly: https://github.com/berstend/puppeteer-extra/issues/209

The goal is to have Puppeteer run every command in a Isolated Worlds to avoid detection scripts being able to monitor execution.

The concept write up is here _(sorry, needed too much detail to include within the issue text)_: https://github.com/prescience-data/harden-puppeteer

The only way I can figure out how to achieve this is by modifying the vanilla Puppeteer files directly in the `node_modules` folder, so hoping someone with experience writing Puppteer-Extra plugins can advise a way to do with with a plugin instead.

Thanks!

**Comment by brunogaspar** [comment]
Have you looked/tried https://github.com/ds300/patch-package ?

**Comment by prescience-data** [comment]
@brunogaspar thanks! Just re-did the concept as a patch, much much easier to follow compared to the previous way!

**Comment by prescience-data** [comment]
My main concern is whether this will somehow mess up plugins like Extra-Stealth by accidently running all of their modifications within the isolated world (ie effectively disabling them)

**Comment by brunogaspar** [comment]
Just a matter of trying, have you tried with the latest Puppeteer?

**Comment by prescience-data** [comment]
Yes, it does work (on 1.19.0, will be updating it 2.1.0) - I'm just trying to think of unknown unknowns etc. Do you know of any ways to test Extra-Stealth features to make sure they are all active?

**Comment by brunogaspar** [comment]
hmm you mean, to determine if the stuff that the stealth plugin does is still being applied? If that's it, i suppose you can try to mimic what the unit tests for the stealth plugin does.

If that's not it, please elaborate a bit more and i'll try to help you out.

**Comment by berstend** [comment]
> My main concern is whether this will somehow mess up plugins like Extra-Stealth by accidently running all of their modifications within the isolated world (ie effectively disabling them)

This should become apparent immediately when using your patches and running `yarn test` in the stealth plugin repo. :-)

Haven't looked more closely at isolated worlds so far but is it similar to what happens in Chrome Extensions and Content Scripts? If so then this would have an effect, as the Puppeteer scripts couldn't access the site's local `window` object (only DOM) without injecting another script in the site.

**Comment by prescience-data** [comment]
> Haven't looked more closely at isolated worlds so far but is it similar to what happens in Chrome Extensions and Content Scripts? If so then this would have an effect, as the Puppeteer scripts couldn't access the site's local `window` object (only DOM) without injecting another script in the site.

Yes it's the same as how the Content Scripts work from memory. 

What I've tried to do is to isolate only the commands sent by the user, meaning the rest of Puppeteer should run normally, but any detection scripts will be unable to monitor your commands, other than to see the outcome in the DOM.

The trade off is that any global libraries you might be expecting to have access to, you'll need to include directly in the script rather than look for them on `window._____`, and naturally that means if you need to interact with the site's custom scripts directly you might not be able to do this (have not tested this though).

**Comment by prescience-data** [comment]
Ok so running the tests in `puppeteer-extra-plugin-stealth` dumps a bunch of these errors with the patch applied:

```
 Rejected promise returned by test. Reason:

  Error {
    message: `Evaluation failed: ReferenceError: fpCollect is not definedΓÉè
        at jquery.js:1:18`,
  }
```

Which would be expected if `fpCollect` is defined outside the isolated world, but the test seems to be testing from "inside" Puppeteer, whereas I think a more accurate test would be inspecting it from "outside", as a detection script would?

**Comment by berstend** [comment]
But don't we intentionally want to run in the same context as the site's JS in order to be able to access and modify it?

Let's make a simpler test case to help understand this:

```js
await page.evaluateOnNewDocument(() => {
  delete Object.getPrototypeOf(navigator).webdriver
})
```

Would this work with your patched files? If not (similar to Content Script isolation in Chrome) we'd need to inject another JS script into the site with the actual payload, which is trivial to detect (MutationObservers, Content Security Policies).

**Comment by prescience-data** [comment]
I don't believe so because this is the `page.evaluateOnNewDocument` function:

```js
  /**
   * @param {Function|string} pageFunction
   * @param {!Array<*>} args
   */
  async evaluateOnNewDocument(pageFunction, ...args) {
    const source = helper.evaluationString(pageFunction, ...args);
    await this._client.send('Page.addScriptToEvaluateOnNewDocument', { source });
  }
```

You can see it is sending the command directly to the `_client` rather than passing through `FrameManager` (which is where the isolated world exists).

The isolated world is set up to catch things like `page.evaluate()` eg:

```js
  /**
   * @param {Function|string} pageFunction
   * @param {!Array<*>} args
   * @return {!Promise<*>}
   */
  async evaluate(pageFunction, ...args) {
    return this._frameManager.mainFrame().evaluate(pageFunction, ...args);
  }
```

Which has been overridden here https://github.com/prescience-data/harden-puppeteer/blob/ba202cc0a422b257c26f023fbaafd41f7ae48157/patches/puppeteer%2B1.19.0.patch#L86 to:

```js
return this._frameManager.isolatedWorld().evaluate(pageFunction, ...args);
```

**Comment by prescience-data** [comment]
> _edit, and then navigating to https://bot.sannysoft.com/ and see if `Webdriver` is missing_

Passed:
![Pass](https://i.snipboard.io/tMSasm.jpg)


Is this what you were expecting for webdriver?

**Comment by prescience-data** [comment]
Ok so I can confirm that the patch works as intended. 

I've made a test for it here that uses Vastel's execution monitoring technique to figure out if the host site has any visibility into the patched context:

Puppeteer Test: https://github.com/prescience-data/puppeteer-botcheck/blob/b6848845b8b5887608784caa2fe7a078db866e9b/Botcheck.js#L45
Host Monitoring Execution: https://github.com/prescience-data/prescience-data.github.io/blob/master/execution-monitor.html

URL of the live test: https://prescience-data.github.io/execution-monitor.html

Here's the differences between unpatched and patched:

![image](https://user-images.githubusercontent.com/65471523/88447898-9a1b4880-ce7b-11ea-89b5-73b1e3cd5332.png)


You can see that the patched version only detects the inserted elements (which was left deliberately unisolated to allow user to inject scripts into the main context (ie all the extra-stealth modifications).

However, anything other than that is running isolated and outside the security scope of any bot detection script. 

Naturally they would be able to observe changes you make to the DOM, but *only* the outcome, not **how** the execution is occurring.
